[SPARK-27045][SQL] SQL tab in UI shows actual SQL instead of callsite in case of SparkSQLDriver

ajithme opened the pull request:
## What changes were proposed in this pull request?

When we run sql in spark via SparkSQLDriver (thrift server, spark-sql), SQL string is siet via ``setJobDescription``. the SparkUI SQL tab must show SQL instead of stacktrace in case ``setJobDescription`` is set which is more useful to end user. Instead it currently shows in description column the callsite shortform which is less useful

![image](https://user-images.githubusercontent.com/22072336/53734682-aaa7d900-3eaa-11e9-957b-0e5006db417e.png)


## How was this patch tested?

Manually:
![image](https://user-images.githubusercontent.com/22072336/53734657-9f54ad80-3eaa-11e9-8dc5-2b38f6970f4e.png)

ajithme committed:
sql UI

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

ajithme commented:
Any suggestions @srowen  @cloud-fan  @vanzin .?

ajithme committed:
import organized

srowen reviewed (approved):
Yeah, this seems much better. I don't know if there are reasons for the current display or implications for changing it, but this seems clearly better.

SparkQA commented:
**[Test build #4588 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4588/testReport)** for PR 23958 at commit [`5dd5bd8`](https://github.com/apache/spark/commit/5dd5bd8362cd741004c9f3c0fcbd26a7b6a66e14).

SparkQA commented:
**[Test build #4588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4588/testReport)** for PR 23958 at commit [`5dd5bd8`](https://github.com/apache/spark/commit/5dd5bd8362cd741004c9f3c0fcbd26a7b6a66e14).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

dongjoon-hyun commented:
Hi, @ajithme . This is an improvement instead of a bug fix. Could you update the title of this PR and JIRA?

dongjoon-hyun reviewed (changes requested):
This is good for UX, but `SparkContext.SPARK_JOB_DESCRIPTION` is a full SQL command string. In case of huge SQL statement, this can increase the size of Spark event log additionally. Usually, the increased Spark event log size is harmful for HDFS NN performance in the clusters. 

I prefer to have a configuration to turn off this or to shorten the SQL string for safety.

cc @gatorsmile and @cloud-fan

viirya commented:
Seems like a good thing to do for UI, but worrying about longer queries might be also not good for users.

cloud-fan commented:
Looks good for UX, but agree with others that we need to do something for super-long queries.

srowen commented:
For the UX, we can always truncate the string.

For event logs, isn't it also important to have the actual query? I understand perf can be an issue but how frequent and how large can queries be vs the value of capturing that? can what's recorded just also be truncated? right now the string here is just a description anyway.

I don't think this should be configurable.

ajithme commented:
i have a small suggestion. Can we provide a config for max length of SQL to be shown in UI.?? (any query beyond it will be truncated in UI, but still be available as a whole in event log)
Please provide your opinion

srowen commented:
I really don't think a config helps here. I think it's just trying to push the question on the user. Just truncate it 'reasonably' in the table and give access to the full string elsewhere. In the logs, is the full query really going to cause a problem?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -77,7 +78,9 @@ object SQLExecution {
             executionId = executionId,
-            description = callSite.shortForm,
+            description = Option(sparkSession.sparkContext
+              .getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION))
```
You know, we should also redact i it. 

Also cc @gengliangwang

ajithme replied:
@gatorsmile any suggestions on how to do it.? how do we identify 'sensitive' part of SQL which we need to hide.?

gatorsmile replied:
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L166-L171 

Please check whether the strings in SparkContext.SPARK_JOB_DESCRIPTION have been redacted. 

Thanks!

ajithme replied:
@gatorsmile i have updated the PR to handle redact of string in SparkContext.SPARK_JOB_DESCRIPTION

ajithme commented:
@srowen apart from size and reading (Example: populating to history server) i do not see any other impact of adding compete query to logs. Please correct me if i am wrong here.

vanzin commented:
SPARK-26103 is something that can happen with large strings; I don't expect SQL queries to get as large as some generated plan strings, though.

ajithme committed:
handle redact of jobdescription

gengliangwang reviewed (approved):
Nice catch!  LGTM.   For long queries:
1. We can add CSS property `word-wrap` as `break-word`  to the description so that it will be consistent with the Jobs page.
2. Or we can only show part of long queries in the SQL table, and show full description in the SQL execution detail page.

Either way is fine to me.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -71,13 +73,21 @@ object SQLExecution {
 
+      val desc = Option(sparkSession.sparkContext
+        .getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION)) match {
+        case value : Some[String] =>
```
Nit: `case Some(value) =>` and then you don't need `.get`.
I think this should be indented another unit, or else move the `getLocalProperty` call into a `val` above. Then the Option-match-Some-None pattern is kind of superfluous. Just check for null.

ajithme replied:
Done

dongjoon-hyun commented:
Yes. @ajithme . Please add two configurations for those maximum length. One for event log and one for the UI.
> Can we provide a config for max length of SQL to be shown in UI?? 

@srowen . For that users, we can get the full query if users set the configuration to `0`. Is it okay for your concern?
> For event logs, isn't it also important to have the actual query?

srowen commented:
Two configs? this seems excessive. Are people going to find, set and maintain these? What's the default? _If_ it's 0, then we still have this potential problem about big query logs and people have to figure out this config exists to change it. If it's nonzero, we aren't logging full info in the logs. This is why I think configs are often just pushing a choice to the user that may be unuseful.

I'm still not clear logging big SQL queries is a problem. Plans can be huge; how big are SQL queries, realistically? kilobytes? I'm not sure there's a real problem here.

dongjoon-hyun commented:
@srowen . I know a few customers which turned off Spark event logs due to the traffic issue. Not every clusters are resourceful in reality. IMO, this PR makes the situation only worse.

srowen commented:
If you turn off event logs, does what gets logged matter?

dongjoon-hyun commented:
If we provide a reasonable configuration to control the degree of size, they will not turn off it. Otherwise, we are going to lose more customers.

We can dump everything more and more in event logs. But, if that makes the whole event log feature off, what's the benefit?

srowen commented:
I'm not sure it follows that this causes people to turn off event logs; it just doesn't seem that big. It's also useful information that isn't being logged now. I wouldn't argue strongly against a flag, if the default is 'no truncation'

cloud-fan commented:
wait, is it an existing problem? how do we deal with super long callsite?

ajithme commented:
@dongjoon-hyun @srowen @gengliangwang @cloud-fan @gatorsmile  So is it acceptable if we introduce two configs
1. **config1** - this will control how much is the threshold of SQL string length beyond which it will be truncated for UI. Defaults to 0 means no truncation
1. **config2** - this will control how much is the threshold of SQL string length beyond which it will be truncated for eventlogs. Defaults to 0 means no truncation
Any user who has problem due to super large queries as mentioned by all the comments, can tune these params accordingly.
Should this address our concerns.?

gengliangwang commented:
For long queries, how about collapsing the ending part and showing it on click. E.g.
![image](https://user-images.githubusercontent.com/1097932/54041959-01434900-4204-11e9-9b43-32eab7dd6aa0.png)


+1 with @srowen . I don't think having a configuration for a web page would be helpful.

dongjoon-hyun commented:
@srowen . Yes. I agree that the default should be `0`.
> I wouldn't argue strongly against a flag, if the default is 'no truncation'

@cloud-fan . In general, for long callSite performance issues, Spark used `withDummyCallSite` as a workaround. However, there is no configuration for event logs, isn't it? BTW, this PR is trying to replace `short callSite` to somewhat unpredictable longer SQL statement. Usually, BI queries are longer than expected. That's the reason we have a longer discussion in this PR.

@ajithme I agree with the following of course. For me, `config2` is a must-have item. For `config1`, please follow other committers' opinion. I'm neutral. (It's good to ignore)
> I think SQL tab primary responsibility should be showing the Entire SQL of what user is trying to run.

ajithme committed:
introduce a truncate conf for truncating super long sqls

ajithme committed:
handle UI comments

ajithme committed:
resolve conflict

ajithme committed:
resolve conflicts

ajithme committed:
static checks

ajithme commented:
@dongjoon-hyun @srowen @gengliangwang @cloud-fan @gatorsmile @gengliangwang 

Thanks for all the inputs. So the new UI looks like this
![image](https://user-images.githubusercontent.com/22072336/54068213-3a190780-4270-11e9-87a0-aa695750bc14.png)


Any query which is exceeding 1 line will be ended by ellipse. User can view the complete query in details. I have updated the PR accordingly. and introduced a new config **spark.sql.event.truncate.length**  please suggest if config name is not satisfactory

One small question about the truncation. Is it ok truncation happen @ creation of SparkListenerSQLExecutionStart.??

ajithme commented:
Hi all, any more suggestions for this PR or are the changes good to go.?

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala:
```diff
@@ -388,14 +388,15 @@ private[ui] class ExecutionPagedTable(
       </span> ++
       <div class="stage-details collapsed">
-        <pre>{execution.details}</pre>
+        <pre>{execution.description}<br></br>{execution.details}</pre>
```
If there is no truncation as default, we can ignore the duplicated `execution.description` here.

ajithme replied:
@gengliangwang that may be hard to do. For example a query may not be truncated but even a browser window resize event, css will truncate the query as per new window size without a server call right.? how to handle then.?

ajithme replied:
@gengliangwang i hope you are talking about css truncation and not the server side truncation.. right.?

gengliangwang replied:
I see. No problem here :)

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1708,6 +1708,13 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
+
+  val SQL_EVENT_TRUNCATE_LENGTH = buildConf("spark.sql.event.truncate.length")
```
Hi, @ajithme .
- Can we move this to `StaticSQLConf` like `spark.sql.ui.retainedExecutions` configuration?
- Naming is always difficult. For config name, I'd suggest `spark.eventLog.sql.maxLength`. Let's get some comment after updating to that.

ajithme replied:
@dongjoon-hyun 
1. Yes. I have moved it to StaticSQLConf 
2. i dint want to mention ``eventLog`` in the configuration because this property will truncate SQL in the ``SparkListenerSQLExecutionStart`` creation itself. ``EventLog`` is just one of the Listener to the event and infact all listeners of this event will get the truncated SQL.  right.?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2151,6 +2158,9 @@ class SQLConf extends Serializable with Logging {
   def setCommandRejectsSparkCoreConfs: Boolean =
     getConf(SQLConf.SET_COMMAND_REJECTS_SPARK_CORE_CONFS)
 
+  def maxSqlForEventLength: Int =
```
In general, function name should be matched to configuration name. Please update the variable name and function name together consistently.

ajithme replied:
i have removed it as we can use StaticSQLConf reference directly

dongjoon-hyun reviewed (commented):
Thank you for updating. Since the configuration is added, there is no blocker for this issue. I left a few comments.

ajithme committed:
move config to StaticSQLConf

ajithme commented:
I have updated after fixing the review comments. Please check

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala:
```diff
@@ -388,14 +388,15 @@ private[ui] class ExecutionPagedTable(
       </span> ++
       <div class="stage-details collapsed">
-        <pre>{execution.details}</pre>
+        <pre>{execution.description}<br></br>{execution.details}</pre>
```
why is this change?

ajithme replied:
@cloud-fan you can see the attached snapshot of after change in UI. With this change we can see SQL and details in the details box ``+details``

ajithme replied:
https://github.com/apache/spark/pull/23958#issuecomment-471156490

cloud-fan replied:
does it introduce changes to non-SQL queries? Sorry I'm not familiar with the frontend part.

ajithme replied:
non sql queries will have callsite.short form instead of sql which is one liner usually

cloud-fan replied:
even without this change, we can still see the full SQL when clicking the `+details`, right? IIUC the full SQL is in `execution.details` not `execution.description`

ajithme replied:
No. You can refer at ``org.apache.spark.sql.execution.ui.SQLAppStatusListener#onExecutionStart`` here you can see description is copied from ``SparkListenerSQLExecutionStart.description`` 

Refer code @ ``org.apache.spark.sql.execution.SQLExecution#withNewExecutionId``
``SparkListenerSQLExecutionStart.description`` maps to ``execution.description`` -- contain full SQL or callsite.shortform
``SparkListenerSQLExecutionStart.details`` maps to ``execution.details `` -- contains callsite.longform

cloud-fan replied:
How can we know the full SQL if it's truncated in `SQLExecution#withNewExecutionId`?

ajithme replied:
We cannot know. There are two truncations here 
1. Backend - Done @ ``SQLExecution#withNewExecutionId`` this is controlled by config and if exceeds threshold string will be truncated before posting ``SparkListenerSQLExecutionStart``.
2. UI - This gets input via ``SparkListenerSQLExecutionStart`` from status store (truncated or may not be truncated) but this may need additional truncation based on user window/browser size. this is done by css - onload or even window resize will trigger this truncation. This is not configurable. This is adjusted as per window size on runtime by css

ajithme replied:
Refer https://github.com/apache/spark/pull/23958#discussion_r264091072

cloud-fan replied:
ah now I got it, thanks for the explanation!

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -71,13 +74,24 @@ object SQLExecution {
+        .getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION)).get
+
+      val truncateLength = sparkSession.sparkContext.conf.get(SQL_EVENT_TRUNCATE_LENGTH)
+      desc = if (desc == null || truncateLength == 0) {
```
let's follow the scala code style
```
val desc = Option(sc.getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION))
  .filter(_ => truncateLength > 0)
  .map { sqlStr =>
    Utils.redact(...)
  }
  .getOrElse(callSite.shortForm)
```

ajithme replied:
Yes, seems better. Done

ajithme committed:
update as per scala style

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -71,13 +74,23 @@ object SQLExecution {
 
+      val truncateLength = sparkSession.sparkContext.conf.get(SQL_EVENT_TRUNCATE_LENGTH)
+
+      val desc = Option(sparkSession.sparkContext
```
nit: `sc` is available here

ajithme replied:
done

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -71,13 +74,23 @@ object SQLExecution {
+        .getLocalProperty(SparkContext.SPARK_JOB_DESCRIPTION))
+        .filter(_ => truncateLength > 0)
+        .map { sqlStr =>
+          Utils.redact(sparkSession.sessionState.conf.stringRedactionPattern, sqlStr)
```
ditto

ajithme replied:
this is SQLConf got via sessionstate.. i do not see any spark context

ajithme committed:
use avaialble sc instead of refering it from session

dongjoon-hyun commented:
@ajithme . Could you update the PR title and description because this PR only work for Spark Thrift Server. Currently, this PR claims too much like the following. However, `spark-shell` is unable to show SQL statements because `setJobDescription` is not called with SQL statement.
> When we run sql in spark ( for example via thrift server), ...

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -71,13 +74,22 @@ object SQLExecution {
+        .filter(_ => truncateLength > 0)
+        .map { sqlStr =>
+          Utils.redact(sparkSession.sessionState.conf.stringRedactionPattern, sqlStr)
+            .substring(0, Math.min(truncateLength, sqlStr.length))
```
Ur, this line is unsafe because it will throw `IndexOutOfBoundsException` if redacted string is smaller than the original. We need to the length of redacted string to call `substring`.

ajithme replied:
Yes, agreed. Done

ajithme committed:
substring based on redacted string

ajithme commented:
@dongjoon-hyun i have updated PR and description. Does now it seems right.?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.sql.internal
 
+import org.apache.spark.sql.internal.SQLConf.buildConf
```
unnecessary import?

ajithme replied:
removed

ajithme committed:
organize imports

cloud-fan commented:
retest this please

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented:
Thanks. @ajithme . It's correct now. This works for STS and spark-sql shell.

dongjoon-hyun reviewed (approved):
+1, LGTM. We need a Jenkins run. It seems that Jenkins farm is out of order now for some reasons. I'll wait for the Jenkins result. Thank you, all.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8803/
Test PASSed.

SparkQA commented:
**[Test build #103378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103378/testReport)** for PR 23958 at commit [`d9da696`](https://github.com/apache/spark/commit/d9da69634a62563539475f4177ec300639fea1c4).

SparkQA commented:
**[Test build #103378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103378/testReport)** for PR 23958 at commit [`d9da696`](https://github.com/apache/spark/commit/d9da69634a62563539475f4177ec300639fea1c4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103378/
Test PASSed.

dongjoon-hyun commented:
Merged to master.
